src/platform_impl/macos/mod.rs

```diff
@@ -1 +1,674 @@
+        self.0.borrow_mut().submenu.as_ref().unwrap().show_context_menu_for_nsview(view)
+    }
+
+    pub fn set_windows_menu_for_nsapp(&self) {
+        unsafe {
+            NSApp().setWindowsMenu_(self.0.borrow().submenu.as_ref().unwrap().ns_menu)
+        }
+    }
+
+    pub fn set_help_menu_for_nsapp(&self) {
+        unsafe {
+            msg_send![NSApp(), setHelpMenu: self.0.borrow().submenu.as_ref().unwrap().ns_menu]
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct MenuItem(Rc<RefCell<MenuChild>>);
+
+impl MenuItem {
+    pub fn new(text: &str, enabled: bool, accelerator: Option<Accelerator>) -> Self {
+        Self(Rc::new(RefCell::new(MenuChild {
+            type_: MenuItemType::Normal,
+            text: strip_mnemonic(text),
+            enabled,
+            id: COUNTER.next(),
+            accelerator,
+            ..Default::default()
+        })))
+    }
+
+    pub fn make_ns_item_for_menu(&self, menu: &Menu) -> id {
+        let mut child = self.0.borrow_mut();
+
+        let ns_menu_item = create_ns_menu_item(&child.text, Some(sel!(fireMenuItemAction:)), &child.accelerator);
+
+        unsafe {
+            let _: () = msg_send![ns_menu_item, setTarget:ns_menu_item];
+            let _: () = msg_send![ns_menu_item, setTag:child.id()];
+
+            // Store a raw pointer to the `MenuChild` as an instance variable on the native menu item
+            let ptr = Box::into_raw(Box::new(&*child));
+            (&mut *ns_menu_item).set_ivar(BLOCK_PTR, ptr as usize);
+
+            if !child.enabled {
+                let () = msg_send![ns_menu_item, setEnabled: NO];
+            }
+        }
+
+        child.ns_menu_items
+            .entry(menu.id)
+            .or_insert(Vec::new())
+            .push(ns_menu_item);
+
+        ns_menu_item
+    }
+
+    pub fn id(&self) -> u32 {
+        self.0.borrow().id()
+    }
+
+    pub fn text(&self) -> String {
+        self.0.borrow().text()
+    }
+
+    pub fn set_text(&self, text: &str) {
+        self.0.borrow_mut().set_text(text)
+    }
+
+    pub fn is_enabled(&self) -> bool {
+        self.0.borrow().is_enabled()
+    }
+
+    pub fn set_enabled(&self, enabled: bool) {
+        self.0.borrow_mut().set_enabled(enabled)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct PredefinedMenuItem(Rc<RefCell<MenuChild>>);
+
+impl PredefinedMenuItem {
+    pub fn new(item_type: PredfinedMenuItemType, text: Option<String>) -> Self {
+        let text = strip_mnemonic(text.unwrap_or_else(|| match item_type {
+            PredfinedMenuItemType::About(_) => format!("About {}", unsafe { app_name_string() }.unwrap_or_default()).trim().to_string(),
+            PredfinedMenuItemType::Hide => format!("Hide {}", unsafe { app_name_string() }.unwrap_or_default()).trim().to_string(),
+            PredfinedMenuItemType::Quit => format!("Quit {}", unsafe { app_name_string() }.unwrap_or_default()).trim().to_string(),
+            _ => item_type.text().to_string(),
+        }));
```

Comment:
@amrbashir I'm not sure about the approach I've taken here because i18n will be hard / impossible.

I think a better approach would be to add placeholders in the output of [`PredefinedMenuItem::text()`](https://github.com/amrbashir/muda/pull/19/files#diff-ade96ad6ce10818697881295338193a079d0df5ebf874621467084a423951b9bR39) – maybe something like `[APP_NAME]` could be replaced by the actual app name (only on macOS… at least for now?)

That way localized text provided by the user could contain the same placeholders.

What do you think?

Reply:
That's not necessary. If they want to do i18n for a predefined item, they need to also provide the app name if they want to display it.

Reply:
Ok, I'll leave it as is then, at least for now.